File: orders-protocol/src/main/scala/bookingtour/protocols/orders/api/PlannerAccommodationItem.scala
package bookingtour.protocols.orders.api

import java.time.LocalDateTime

import bookingtour.protocols.core._
import bookingtour.protocols.core.newtypes.quantities.{Nights, PropertyStar}
import bookingtour.protocols.core.newtypes.values.{BoardingLabel, RoomCategoryLabel, RoomTypeLabel, TariffLabel}
import bookingtour.protocols.core.values.{Amount, Ranges}
import bookingtour.protocols.orders.newTypes.PlannerAccommodationItemId
import bookingtour.protocols.parties.newTypes.CustomerGroupId
import bookingtour.protocols.properties.newTypes.PropertyId
import bookingtour.protocols.property.prices.newTypes.{PriceUnitId, TariffId}
import cats.instances.all._
import derevo.cats.order
import derevo.circe.{decoder, encoder}
import derevo.derive

/**
  * © Alexey Toroshchin 2020.
  */
@derive(encoder, decoder, order)
final case class PlannerAccommodationItem(
    id: PlannerAccommodationItemId,
    created: LocalDateTime,
    updated: LocalDateTime,
    dates: Ranges.Dates,
    propertyId: PropertyId,
    propertyName: String,
    propertyStar: PropertyStar,
    roomPriceUnitId: PriceUnitId,
    roomType: RoomTypeLabel,
    roomCategory: RoomCategoryLabel,
    boarding: BoardingLabel,
    tariffId: TariffId,
    tariff: TariffLabel,
    groupId: CustomerGroupId,
    nights: Nights,
    price: Amount,
    discount: Option[Amount],
    markAsBest: Boolean,
    markAsDelete: Boolean
)

File: customer-api-spa/types/planner/PlannerAccommodationItem.ts
import { Amount } from '../bookings/Amount'
import { DateRange, formatLocalDateTime } from '../DateRange'
import { v4 as uuid } from 'uuid'
import { Uuid } from '../basic/Uuid'
import { Nullable } from '../Nullable'

export type PlannerAccommodationItemId = string

export interface PlannerAccommodationItem {
  id: PlannerAccommodationItemId;
  created: string;
  updated: string;
  dates: DateRange;
  propertyId: Uuid;
  propertyName: string;
  propertyStar: number;
  roomPriceUnitId: Uuid;
  roomType: string;
  roomCategory: string;
  boarding: string;
  tariffId: Uuid;
  tariff: string;
  groupId: Uuid;
  nights: number;
  price: Amount;
  discount: Nullable<Amount>;
  markAsBest: boolean;
  markAsDelete: boolean;
}

export function makeItem(
  dates: DateRange,
  propertyId: Uuid,
  propertyName: string,
  propertyStar: number,
  roomPriceUnitId: Uuid,
  roomType: string,
  roomCategory: string,
  boarding: string,
  tariffId: Uuid,
  tariff: string,
  groupId: string,
  nights: number,
  amount: Amount,
  discount: Nullable<Amount>
): PlannerAccommodationItem {
  return {
    id: uuid(),
    created: formatLocalDateTime(new Date()),
    updated: formatLocalDateTime(new Date()),
    markAsBest: false,
    markAsDelete: false,
    dates,
    propertyId,
    propertyName,
    propertyStar,
    roomPriceUnitId,
    roomType,
    roomCategory,
    boarding,
    tariffId,
    tariff,
    groupId,
    nights,
    price: amount,
    discount
  }
}
